plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java

```diff
@@ -90,83 +89,107 @@ public BigQueryMetadata(BigQueryClient bigQueryClient, BigQueryConfig config)
                 .collect(toImmutableList());
     }
 
     @Override
     public List<SchemaTableName> listTables(ConnectorSession session, Optional<String> schemaName)
     {
         log.debug("listTables(session=%s, schemaName=%s)", session, schemaName);
-        return listTablesWithTypes(session, schemaName, TABLE);
-    }
-
-    @Override
-    public List<SchemaTableName> listViews(ConnectorSession session, Optional<String> schemaName)
-    {
-        log.debug("listViews(session=%s, schemaName=%s)", session, schemaName);
-        return listTablesWithTypes(session, schemaName, VIEW);
-    }
-
-    private List<SchemaTableName> listTablesWithTypes(ConnectorSession session, Optional<String> schemaName, TableDefinition.Type... types)
-    {
-        if (schemaName.isPresent() && schemaName.get().equalsIgnoreCase(INFORMATION_SCHEMA)) {
+        // filter ambiguous schemas
+        Optional<String> remoteSchema = schemaName.flatMap(schema -> bigQueryClient.toRemoteDataset(projectId, schema)
+                .filter(dataset -> !dataset.isAmbiguous())
+                .map(RemoteDatabaseObject::getOnlyRemoteName));
+        if (remoteSchema.isPresent() && remoteSchema.get().equalsIgnoreCase(INFORMATION_SCHEMA)) {
             return ImmutableList.of();
         }
-        Set<String> schemaNames = schemaName.map(ImmutableSet::of)
-                .orElseGet(() -> ImmutableSet.copyOf(listSchemaNames(session)));
 
-        ImmutableList.Builder<SchemaTableName> tableNames = ImmutableList.builder();
-        for (String datasetId : schemaNames) {
-            for (Table table : bigQueryClient.listTables(DatasetId.of(projectId, datasetId), types)) {
-                tableNames.add(new SchemaTableName(datasetId, table.getTableId().getTable()));
+        Set<String> remoteSchemaNames = remoteSchema.map(ImmutableSet::of)
+                .orElseGet(() -> ImmutableSet.copyOf(listRemoteSchemaNames()));
+
+        ImmutableSet.Builder<SchemaTableName> tableNames = ImmutableSet.builder();
+        for (String remoteSchemaName : remoteSchemaNames) {
+            for (Table table : bigQueryClient.listTables(DatasetId.of(projectId, remoteSchemaName), TABLE, VIEW)) {
+                // filter ambiguous tables
+                boolean isAmbiguous = bigQueryClient.toRemoteTable(projectId, remoteSchemaName, table.getTableId().getTable().toLowerCase(ENGLISH))
+                        .filter(RemoteDatabaseObject::isAmbiguous)
+                        .isPresent();
+                if (!isAmbiguous) {
+                    tableNames.add(new SchemaTableName(table.getTableId().getDataset(), table.getTableId().getTable()));
+                }
+                else {
+                    log.debug("Filtered out [%s.%s] from list of tables due to ambiguous name", remoteSchemaName, table.getTableId().getTable());
+                }
             }
         }
-        return tableNames.build();
+        return tableNames.build().asList();
```

Comment:
can it be ImmutableList.builder() from the start?

Reply:
It can be now that we are removing the duplicates in the loop.

Thanks.